File: VIO/oakd_euroc.py
#!/usr/bin/env python3

# Chobits' script to capture images and IMU from oak-d to follow EuRoC format
import cv2
import depthai as dai
import time
import math
import csv
import pathlib
import numpy as np
import threading, time

gogogo = True
l_img = None
l_ts = 0
r_img = None
r_ts = 0

# path2dataset='/home/ludofw/Data/Drones'
path2dataset='/mnt/Data_3TB/Data/Datasets/Kalibr/oakd_lite_IMU_cam'

def save_png():
    global l_img
    global r_img
    while gogogo:
        if l_img is not None:
            cv2.imwrite(f"{path2dataset}/oakd_lite/mav0/cam0/data/{l_ts}.png", l_img)
            l_img = None
        if r_img is not None:
            cv2.imwrite(f"{path2dataset}/oakd_lite/mav0/cam1/data/{r_ts}.png", r_img)
            r_img = None
        time.sleep(0.0001)

pathlib.Path(path2dataset+"/oakd_lite/mav0/imu0").mkdir(parents=True, exist_ok=True)
pathlib.Path(path2dataset+"/oakd_lite/mav0/cam0/data").mkdir(parents=True, exist_ok=True)
pathlib.Path(path2dataset+"/oakd_lite/mav0/cam1/data").mkdir(parents=True, exist_ok=True)

fs = cv2.FileStorage("q250_imu_cali.yml", cv2.FILE_STORAGE_READ)
acc_misalign = fs.getNode("acc_misalign").mat()
acc_bias = fs.getNode("acc_bias").mat()
acc_scale = fs.getNode("acc_scale").mat()
gyro_misalign = fs.getNode("gyro_misalign").mat()
gyro_scale = fs.getNode("gyro_scale").mat()
gyro_bias = fs.getNode("gyro_bias").mat()
acc_cor =  acc_misalign @ acc_scale
gyro_cor = gyro_misalign @ gyro_scale
fs.release()

# Create pipeline
pipeline = dai.Pipeline()

# Define sources and outputs
imu = pipeline.create(dai.node.IMU)
xout_imu = pipeline.create(dai.node.XLinkOut)
monoLeft = pipeline.create(dai.node.MonoCamera)
xoutLeft = pipeline.create(dai.node.XLinkOut)
monoRight = pipeline.create(dai.node.MonoCamera)
xoutRight = pipeline.create(dai.node.XLinkOut)

xout_imu.setStreamName("imu")
xoutLeft.setStreamName("left")
xoutRight.setStreamName("right")

imu.enableIMUSensor(dai.IMUSensor.ACCELEROMETER_RAW, 200)
imu.enableIMUSensor(dai.IMUSensor.GYROSCOPE_RAW, 200)
# it's recommended to set both setBatchReportThreshold and setMaxBatchReports to 20 when integrating in a pipeline with a lot of input/output connections
# above this threshold packets will be sent in batch of X, if the host is not blocked and USB bandwidth is available
imu.setBatchReportThreshold(5)
# maximum number of IMU packets in a batch, if it's reached device will block sending until host can receive it
# if lower or equal to batchReportThreshold then the sending is always blocking on device
# useful to reduce device's CPU load  and number of lost packets, if CPU load is high on device side due to multiple nodes
imu.setMaxBatchReports(10)
monoLeft.setCamera("left")
monoLeft.setResolution(dai.MonoCameraProperties.SensorResolution.THE_480_P)
monoLeft.setFps(20)
monoRight.setCamera("right")
monoRight.setResolution(dai.MonoCameraProperties.SensorResolution.THE_480_P)
monoRight.setFps(20)

# Link plugins IMU -> XLINK
imu.out.link(xout_imu.input)
monoLeft.out.link(xoutLeft.input)
monoRight.out.link(xoutRight.input)

t_png = threading.Thread(target=save_png)
t_png.start()

# Pipeline is defined, now we can connect to the device
with dai.Device(pipeline) as device, open(path2dataset+'/oakd_lite/mav0/imu0/data.csv', 'w') as imu_file, open(path2dataset+'/oakd_lite/mav0/cam0/data.csv', 'w') as cam0_file, open(path2dataset+'/oakd_lite/mav0/cam1/data.csv', 'w') as cam1_file:
    print("capture started...")
    imu_writer = csv.writer(imu_file)
    cam0_writer = csv.writer(cam0_file)
    cam1_writer = csv.writer(cam1_file)
    # Output queue for imu bulk packets
    imuQueue = device.getOutputQueue(name="imu", maxSize=200, blocking=False)
    qLeft = device.getOutputQueue(name="left", maxSize=4, blocking=False)
    qRight = device.getOutputQueue(name="right", maxSize=4, blocking=False)
    try:
        while True:
            queueName = device.getQueueEvent()
            if queueName == "imu":
                imuData = imuQueue.get()
                imuPackets = imuData.packets
                for imuPacket in imuPackets:
                    acceleroValues = imuPacket.acceleroMeter
                    gyroValues = imuPacket.gyroscope
                    acc_cali = acc_cor @ (np.array([[acceleroValues.x], [acceleroValues.y], [acceleroValues.z]]) - acc_bias)
                    gyro_cali = gyro_cor @ (np.array([[gyroValues.x], [gyroValues.y], [gyroValues.z]]) - gyro_bias)
                    # align with cam axis
                    imu_writer.writerow((int(acceleroValues.getTimestampDevice().total_seconds()*1e9), gyro_cali[0, 0], -gyro_cali[1, 0], -gyro_cali[2, 0], acc_cali[0, 0], -acc_cali[1, 0], -acc_cali[2, 0]))
            elif queueName == "left":
                inLeft = qLeft.get()
                l_ts = int(inLeft.getTimestampDevice().total_seconds()*1e9)
                l_img = inLeft.getFrame()
                # l_img = inLeft.getCvFrame()
                cam0_writer.writerow((l_ts, f"{l_ts}.png"))
                # cv2.imshow('left', l_img)
                # cv2.waitKey(10)
            elif queueName == "right":
                inRight = qRight.get()
                r_ts = int(inRight.getTimestampDevice().total_seconds()*1e9)
                r_img = inRight.getFrame()
                cam1_writer.writerow((r_ts, f"{r_ts}.png"))
    except KeyboardInterrupt:
        print("ctrl_c")
gogogo = False
t_png.join()
print("bye")

File: VIO/cpp/oakd_euroc.cpp
#include <chrono>
#include <iostream>
#include <fstream>
#include <filesystem>
#include <csignal>
#include <atomic>

#include <opencv2/calib3d.hpp>
// Includes common necessary includes for development using depthai library
#include "depthai/depthai.hpp"

std::atomic<bool> go_flag(true);

void signalHandler(int signum) {
    go_flag = false;
}

int main(int argc, char** argv) {
    using namespace std::chrono;

    std::signal(SIGINT, signalHandler); // Register handler for Ctrl-C

    cv::FileStorage imu_yml;
    imu_yml.open(argv[1], cv::FileStorage::READ);
    cv::Mat acc_mis_align, acc_scale, acc_bias;
    imu_yml["acc_misalign"] >> acc_mis_align;
    imu_yml["acc_scale"] >> acc_scale;
    imu_yml["acc_bias"] >> acc_bias;
    cv::Mat acc_cor = acc_mis_align * acc_scale;
    cv::Mat gyro_mis_align, gyro_scale, gyro_bias;
    imu_yml["gyro_misalign"] >> gyro_mis_align;
    imu_yml["gyro_scale"] >> gyro_scale;
    imu_yml["gyro_bias"] >> gyro_bias;
    cv::Mat gyro_cor = gyro_mis_align * gyro_scale;
    imu_yml.release();

    // Create pipeline
    dai::Pipeline pipeline;

    // Define source and output
    auto monoLeft = pipeline.create<dai::node::MonoCamera>();
    auto xoutLeft = pipeline.create<dai::node::XLinkOut>();
    auto monoRight = pipeline.create<dai::node::MonoCamera>();
    auto xoutRight = pipeline.create<dai::node::XLinkOut>();
    auto imu = pipeline.create<dai::node::IMU>();
    auto xout_imu = pipeline.create<dai::node::XLinkOut>();

    xoutLeft->setStreamName("left");
    xoutRight->setStreamName("right");
    xout_imu->setStreamName("imu");

    // Properties
    monoLeft->setCamera("left");
    monoLeft->setResolution(dai::MonoCameraProperties::SensorResolution::THE_480_P);
    monoLeft->setFps(20);
    monoRight->setCamera("right");
    monoRight->setResolution(dai::MonoCameraProperties::SensorResolution::THE_480_P);
    monoRight->setFps(20);
    imu->enableIMUSensor(dai::IMUSensor::ACCELEROMETER_RAW, 200);
    imu->enableIMUSensor(dai::IMUSensor::GYROSCOPE_RAW, 200);
    imu->setBatchReportThreshold(5);
    imu->setMaxBatchReports(10);

    // Linking
    monoLeft->out.link(xoutLeft->input);
    monoRight->out.link(xoutRight->input);
    imu->out.link(xout_imu->input);

    // Connect to device and start pipeline
    dai::Device device(pipeline);

    std::cout << "Usb speed: " << device.getUsbSpeed() << "\n";
    std::cout << "Device name: " << device.getDeviceName() << " Product name: " << device.getProductName() << "\n";

    // Output queue will be used to get the grayscale frames from the output defined above
    auto qLeft = device.getOutputQueue("left", 4, false);
    auto qRight = device.getOutputQueue("right", 4, false);
    auto imuQueue = device.getOutputQueue("imu", 200, false);

    std::string home = "oakd_lite/mav0";
    std::filesystem::create_directories(std::filesystem::path(home + "/cam0/data"));
    std::filesystem::create_directories(std::filesystem::path(home + "/cam1/data"));
    std::filesystem::create_directories(std::filesystem::path(home + "/imu0"));
    std::ofstream imu_file(home + "/imu0/data.csv");
    imu_file << std::setprecision(std::numeric_limits<float>::max_digits10);
    std::ofstream cam0_file(home + "/cam0/data.csv");
    cam0_file << std::setprecision(std::numeric_limits<float>::max_digits10);
    std::ofstream cam1_file(home + "/cam1/data.csv");
    cam1_file << std::setprecision(std::numeric_limits<float>::max_digits10);

    while(go_flag) {
        auto q_name = device.getQueueEvent();
        if (q_name == "left") {
            auto frame = qLeft->get<dai::ImgFrame>();
            int64_t ns = std::chrono::duration_cast<std::chrono::nanoseconds>(frame->getTimestampDevice().time_since_epoch()).count();
            cv::imwrite(home + "/cam0/data/" + std::to_string(ns) + ".png", frame->getCvFrame());
            cam0_file << ns << "," << ns << ".png\n";
        } else if (q_name == "right") {
            auto frame = qRight->get<dai::ImgFrame>();
            int64_t ns = std::chrono::duration_cast<std::chrono::nanoseconds>(frame->getTimestampDevice().time_since_epoch()).count();
            cv::imwrite(home + "/cam1/data/" + std::to_string(ns) + ".png", frame->getCvFrame());
            cam1_file << ns << "," << ns << ".png\n";
        } else if (q_name == "imu") {
            auto imuData = imuQueue->get<dai::IMUData>();
            auto imuPackets = imuData->packets;
            for(const auto& imuPacket : imuPackets) {
                auto& acc = imuPacket.acceleroMeter;
                auto& gyro = imuPacket.gyroscope;
                cv::Mat acc_raw = (cv::Mat_<double>(3,1) << acc.x, acc.y, acc.z);
                cv::Mat1d acc_cali = acc_cor * (acc_raw - acc_bias);
                cv::Mat gyro_raw = (cv::Mat_<double>(3,1) << gyro.x, gyro.y, gyro.z);
                cv::Mat1d gyro_cali = gyro_cor * (gyro_raw - gyro_bias);
                auto ns = std::chrono::duration_cast<std::chrono::nanoseconds>(acc.getTimestampDevice().time_since_epoch()).count();
                imu_file << ns << "," << gyro_cali(0,0) << "," << -gyro_cali(1,0) << "," << -gyro_cali(2,0) << "," << acc_cali(0,0) << "," << -acc_cali(1,0) << "," << -acc_cali(2,0) << "\n";
            }
        }
    }
    imu_file.close();
    cam0_file.close();
    cam1_file.close();
    std::cout<<"bye\n";
    return 0;
}
